### PR TITLE
Use custom script wrapper for all km tests to capture user, km dumps on hangs

### LINKED ISSUE #3162: Capture user and kernel dumps for all kernel mode tests for in cases of run-time hangs
### Describe the bug

The kernel tests are currently being run directly in the CI/CD pipeline. If these tests hang for any reason, the github timeout (90 minutes) kicks in and cancels the job.  There is no provision for killing the test process and harvesting the kernel and user dumps in this event.

All kernel mode tests should be run such that if the test hangs (does not complete within the specified timeout) we capture user and kernel mode dumps.

The `procdump64.exe` and `notmyfault64.exe` tools from the SysInternals suite can be used to capture the user and kernel mode dumps respectively.

### OS information

_No response_

### Steps taken to reproduce bug

local testing, CI/CD run

### Expected behavior

If the test hangs, user and kernel mode dumps should be created and the test terminated thereafter.

### Actual outcome

The job is cancelled by github after it's timeout (90 minutes) and no dump is created.

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by dv-msft**
## Description

This PR runs all kernel mode tests using the `Run-Self-Hosted-Runner-Test.ps1` wrapper script.  This script watches for a 'hung' test and on expiration of the specified timeout, creates a user and kernel mode dump for the test.

## Testing

(This is a testing enhancement.)
Verified creation of user and kernel mode dumps on timeout expiry by forcing timeouts < test runtime.

Sample output (from `run-self-hosted-runner-test.ps1`) on determining a failing test based on examination of test process exit code (_test failure simulated by running the test with the in-kernel `ebpf-for-windows` drivers in a 'stopped' state_):

```
[16:04:58] :: Command               : .\xdp_tests.exe
[16:04:58] :: Arguments             : xdp_reflect_test --remote-ip 192.168.1.11
[16:04:58] :: Test Hang Timeout     : 3600
[16:04:58] :: User mode dump Folder : C:\Dumps
[16:04:58] :: Kernel dump needed    : True
[16:04:59] :: Verifying ProcDump64.exe presence in C:\eBPF...
[16:04:59] :: ProcDump64.exe location: C:\eBPF\procdump64.exe
[16:04:59] ::

[16:04:59] :: Verifying NotMyFault64.exe presence in C:\eBPF...
[16:04:59] :: NotMyFault64.exe location: C:\eBPF\notmyfault64.exe
[16:04:59] ::

[16:04:59] ::

[16:04:59] :: Current kernel dump configuration:

[16:04:59] ::   @{AutoReboot=1
[16:04:59] ::   CrashDumpEnabled=7
[16:04:59] ::   DisableEmoticon=1
[16:04:59] ::   DumpFile=C:\Windows\MEMORY.DMP
[16:04:59] ::   LogEvent=1
[16:04:59] ::   MinidumpDir=C:\Windows\Minidump
[16:04:59] ::   MinidumpsCount=5
[16:04:59] ::   Overwrite=1
[16:04:59] ::   PSPath=Microsoft.PowerShell.Core\Registry::HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\CrashControl
[16:04:59] ::   PSParentPath=Microsoft.PowerShell.Core\Registry::HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control
[16:04:59] ::   PSChildName=CrashControl
[16:04:59] ::   PSDrive=HKLM
[16:04:59] ::   PSProvider=Microsoft.PowerShell.Core\Registry}
[16:04:59] ::

[16:05:01] :: Available System disk space (Before test start): 26.62 GB
[16:05:01] ::


[16:05:01] :: Staring Test command: C:\eBPF\.\xdp_tests.exe xdp_reflect_test --remote-ip 192.168.1.11
[16:05:01] :: Test hang timeout: 3600 (seconds)
[16:05:01] ::

[16:05:03] ::


[16:05:03] :: Test Program output:

[16:05:03] :: Filters: "xdp_reflect_test"
[16:05:03] :: Randomness seeded to: 1082229669
[16:05:03] :: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[16:05:03] :: xdp_tests.exe is a Catch2 v3.5.3 host application.
[16:05:03] :: Run with -? for options
[16:05:03] :: -------------------------------------------------------------------------------
[16:05:03] :: xdp_reflect_test
[16:05:03] :: -------------------------------------------------------------------------------
[16:05:03] :: D:\wrk\ebpf-for-windows\tests\xdp\xdp_tests.cpp(73)
[16:05:03] :: ...............................................................................
[16:05:03] :: D:\wrk\ebpf-for-windows\tests\libs\util\socket_helper.cpp(214): FAILED:
[16:05:03] :: explicitly with message:
[16:05:03] ::   Receiver socket did not receive any message in 1 second.
[16:05:03] :: ===============================================================================
[16:05:03] :: test cases: 1 | 1 failed
[16:05:03] :: assertions: 1 | 1 failed
[16:05:03] ::

[16:05:03] :: *** ERROR *** C:\eBPF\Run-Self-Hosted-Runner-Test.ps1: C:\eBPF\.\xdp_tests.exe failed.
*** ERROR *** C:\eBPF\Run-Self-Hosted-Runner-Test.ps1: C:\eBPF\.\xdp_tests.exe failed.
At D:\wrk\ebpf-for-windows\x64\Release\vm_run_tests.psm1:255 char:5
+     Invoke-Command -VMName $VM -Credential $TestCredential -ScriptBlo ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (*** ERROR *** C...sts.exe failed.:String) [], RuntimeException
    + FullyQualifiedErrorId : *** ERROR *** C:\eBPF\Run-Self-Hosted-Runner-Test.ps1: C:\eBPF\.\xdp_tests.exe failed.
    + PSComputerName        : CICD-WS19

```

## Documentation

No doc changes.

## Installation

No installer impact.

Fixes #3162

### CODE REVIEW

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    Write-Log "Current kernel dump configuration:`n"
+    $KernelDumpInfo = Get-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\CrashControl'
+    $Lines = $KernelDumpInfo -split '; '
+    foreach($Line in $Lines) {
+        Write-Log "`t$Line"
+    }
+    Write-Log "`n"
+}
+
+$BeforeTestFreeSpaceGB = (((Get-Volume C).SizeRemaining) / 1GB).ToString("F2")
```
[the remark this answers is not in the document]
Comment (dv-msft) [reply scripts/Run-Self-Hosted-Runner-Test.ps1]: fixed.

**Review by Alan-Jowett — APPROVED** [review]

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 2]
```diff
@@ -0,0 +1,265 @@
+# Copyright (c) Microsoft Corporation
+# SPDX-License-Identifier: MIT
+#
+# This script executes the provided test command, waits for <TestHangTimeout> (in seconds) and then triggers user and
+# kernel dumps if the test process is still ruunning (This is typically the case when the test process is hung for some
```
Comment (nibanks) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:5]:
```suggestion
# kernel dumps if the test process is still running (This is typically the case when the test process is hung for some
```

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+# reason).
+#
+# The user mode dump is created using 'procdump64.exe' and the kernel dump using 'notmyfault64.exe' utilities
```
Comment (nibanks) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:8]: Why not livedump64.exe?

File: scripts/run_driver_tests.psm1 [hunk 4]
```diff
@@ -67,178 +69,196 @@ function Invoke-Test
...
+    $TestRunScript = ".\Run-Self-Hosted-Runner-Test.ps1"
+    & $TestRunScript `
+        -TestCommand $TestName `
+        -TestArguments $JoinedArgumentsList `
+        -TestHangTimeout $TestHangTimeout `
+        -UserModeDumpFolder $UserModeDumpFolder `
+        -NeedKernelDump $True `
+        -Verbose
 
     Write-Log "$TestName Passed" -ForegroundColor Green
```
Comment (rectified95) [inline scripts/run_driver_tests.psm1:82]: Make sure you've checked test failures are correctly detected, eg. and a failing assert in several places in xdp_test and connect_redirect_test

File: scripts/run_driver_tests.psm1 [hunk 5]
```diff
@@ -67,178 +69,196 @@ function Invoke-Test
...
+          [parameter(Mandatory = $false)][bool] $Coverage = $false,
+          [parameter(Mandatory = $false)][int] $TestHangTimeout = 3600,
+          [parameter(Mandatory = $false)][string] $UserModeDumpFolder = "C:\Dumps"
+    )
+
 
     pushd $WorkingDirectory
     $env:EBPF_ENABLE_WER_REPORT = "yes"
 
-    try {
```
Comment (rectified95) [inline scripts/run_driver_tests.psm1:108]: why is the try0catch no longer necessary?

File: scripts/Run-Self-Hosted-Runner-Test.ps1 (hunk 3 above)
Comment (dv-msft) [reply scripts/Run-Self-Hosted-Runner-Test.ps1:8]: A live dump would be useful if the hang was in the user mode components of the test.  OTOH, if the hang is in the ebpf kernel components, there's no point in continuing further tests.

I guess determining the location of the hang _might_ be possible, I'm not sure what this additional complexity buys us in our current ci/cd scenario.

File: scripts/vm_run_tests.psm1 [hunk 6]
```diff
@@ -448,8 +598,8 @@ function Invoke-ConnectRedirectTestsOnVM
         }
     }
 
-    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($StandardUserPassword)
-    $UnsecurePassword = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($($StandardUserPassword))
```
Comment (rectified95) [inline scripts/vm_run_tests.psm1:601]: nit, why is this change needed?

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+# While at it, enable EULA for all SysInternals tools.
+REG ADD HKCU\Software\Sysinternals /v EulaAccepted /t REG_DWORD /d 1 /f | Out-Null
+
+# The following 'Set-ItemProperty' command enables a full memory dump.
+# Not enabled yet as it needs a VM with an explicitly created page file of at least (physical_memory + 1MB) in size.
+# The default value of the 'CrashDumpEnabled' key is 7 ('automatic' sizing of dump file size (system determined)).
+# https://learn.microsoft.com/en-us/troubleshoot/windows-server/performance/memory-dump-file-options
+#
+# Set-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\CrashControl' -Name 'CrashDumpEnabled' -Value 1
```
Comment (rectified95) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:128]: A full memory dump might not al;ways be needed, we could later add some conditional logic to select it

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    # First, generate a user mode dump.
+    $UserModeDumpFileName = "$($TestCommand)_$(Get-Date -Format 'yyyy-MM-dd_HH-mm-ss').dmp"
+    $UserModeDumpFilePath = Join-Path $UserModeDumpFolder $UserModeDumpFileName
+
+    if ($VerbosePreference -eq 'Continue') {
+        Write-Log "User mode Dumpfile name: $UserModeDumpFileName"
+        Write-Log "User mode Dumpfile Path: $UserModeDumpFilePath"
+    }
+
+    # Get the available free space at this point in case the test creates files of it's own.
```
Comment (rectified95) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:180]: nit: "files of its own" or simpler "its own files"

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    Write-Log "Waiting for user mode dump to complete..."
+    $ProcDumpProcess.WaitForExit()
+
+    # Get procdump64.exe's exit code.
+    $ExitCode = $($ProcDumpProcess.ExitCode)
+    Write-Log "$ProcDumpBinaryPath completed with exit code: $ExitCode`n"
+
+    # Flush disk buffers to ensure user mode dump data is completely written to disk.
+    Write-Log "User mode dump completed. Flushing disk buffers..."
+    Write-VolumeCache -DriveLetter C
```
Comment (rectified95) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:201]: OOC, have you seen errors related to this?

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        $TestProcess.Kill()
+        ThrowWithErrorMessage `
+            -ErrorMessage "*** ERROR *** $($PSCommandPath): kernel mode dump creation FAILED"
+    } else {
+        Write-Log "`n`n"
+        Write-Log "Kernel dump not needed, killing test process $($TestProcess.ProcessName)..."
+        $TestProcess.Kill()
+        Write-Log "`n`n"
+
+        $ErrorMessage = "*** ERROR *** $($PSCommandPath): $FullTestCommandSpec.exe exceeded test hang timout of " +
```
Comment (rectified95) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:235]: this message seems a duplicate of line 169.
Can you include the sample new output in the PR description?

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+            "$TestHangTimeout seconds"
+        ThrowWithErrorMessage -ErrorMessage $ErrorMessage
+    }
+} else {
+
+    # Ensure the process has completely exited.
+    $TestProcess.WaitForExit()
```
Comment (rectified95) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:242]: don't we reach this else block if the wait has already ended?

File: scripts/Run-Self-Hosted-Runner-Test.ps1 (hunk 7 above)
Comment (dv-msft) [reply scripts/Run-Self-Hosted-Runner-Test.ps1:128]: I disagree.  In my experience it is always better to have the full dump, especially in post-mortem scenarios where the root cause is not immediately obvious and having all possible kernel info comes in handy.

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        Write-Log "Test Program output:`n"
+        foreach($Line in $TestStandardOutput) {
+            if ($Line -ne $Null) {
+                Write-Log $Line
+            }
+        }
+    }
+    Write-Log "`n"
+
+    if ($($TestProcess.ExitCode) -ne 0) {
```
Comment (rectified95) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:257]: Looks like this does the failure detection now, but to make sure we don't have false positive in the CI, please include screenshots of succesfuly detected test failures.

File: scripts/run_driver_tests.psm1 (hunk 4 above)
Comment (dv-msft) [reply scripts/run_driver_tests.psm1:82]: Not sure I understand this comment.  Could you please clarify with an example (if possible)?

File: scripts/run_driver_tests.psm1 (hunk 5 above)
Comment (dv-msft) [reply scripts/run_driver_tests.psm1:108]: The new test runner script throws exceptions on errors and _not_ catching those ensures that the test is marked as 'failed'.  If there's another way of ensuring that failing tests are marked as 'failed' with the try in place, I'd be happy to consider it.  I'm still ramping up on my PS-fu so any pointers would be definitely appreciated.

File: scripts/vm_run_tests.psm1 (hunk 6 above)
Comment (dv-msft) [reply scripts/vm_run_tests.psm1:601]: This seems more of a gh formatting issue? I only renamed the `$UnsecurePassword` to `$InsecurePassword`.

File: scripts/Run-Self-Hosted-Runner-Test.ps1 (hunk 10 above)
Comment (dv-msft) [reply scripts/Run-Self-Hosted-Runner-Test.ps1:235]: We get here only if the test timeout expires _AND_ the kernel dump is disabled (a very rare case, IMO).  We still need to `throw` with a suitable error message.  The slight duplication here makes sense in this very specific (and extremely rare) scenario.

File: scripts/Run-Self-Hosted-Runner-Test.ps1 (hunk 8 above)
Comment (dv-msft) [reply scripts/Run-Self-Hosted-Runner-Test.ps1]: fixed.

File: scripts/run_driver_tests.psm1 (hunk 4 above)
Comment (dv-msft) [reply scripts/run_driver_tests.psm1:82]: Ah. I see what you mean.  So long as the tests themselves report a non-zero exit code in case of such failures, the failures are correctly detected.

File: scripts/Run-Self-Hosted-Runner-Test.ps1 [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+{
+    param(
+        [Parameter(Mandatory = $True)] [string] $ToolName
+    )
+
+    $ToolLocationPath = Get-ChildItem -Path "$Pwd" `
+        -Recurse -Filter "$ToolName" -ErrorAction SilentlyContinue | Select-Object -First 1
+    if ($ToolLocationPath -eq $null) {
+        $ErrorMessage = "*** ERROR *** $ToolName not found under $Pwd."
+        Write-Log $ErrorMessage
```
Comment (matthewige) [inline scripts/Run-Self-Hosted-Runner-Test.ps1:38]: Should you use your helper function ThrowWithErrorMessage ?

Comment (dv-msft) [reply scripts/Run-Self-Hosted-Runner-Test.ps1]: fixed.

File: scripts/Run-Self-Hosted-Runner-Test.ps1 (hunk 12 above)
Comment (dv-msft) [reply scripts/Run-Self-Hosted-Runner-Test.ps1:254]: PR description updated with output from a test failure.

**Review by gtrevi — APPROVED** [review]
